### PR TITLE
filter_multiline: implement Docker partial_message support

### PR DESCRIPTION
**Pull request by PettitWesley**
Signed-off-by: Wesley Pettit <wppttt@amazon.com>

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [ ] Example configuration file for the change
- [ ] Debug log output from testing the change

- [ ] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by PettitWesley** [comment]
@edsiper there won't be another 1.8 version right? So I just need to submit a PR to master for this instead..

**Comment by patrick-stephens** [comment]
> @edsiper there won't be another 1.8 version right? So I just need to submit a PR to master for this instead..

I think the plan is still some maintenance releases @PettitWesley, I've submitted a few PRs for 1.8 for example.

**Comment by edsiper** [comment]
multiline core feature already supports partial docker messages, what is different on this PR ?

**Review by matthewfala — COMMENTED** [review]
Looks good! Added some comments for clairification

File: plugins/filter_multiline/ml.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml.c:574]: The flush timer can't help on shutdown. I need another solution for that (which will come in a separate PR and I will do a separate write up in an issue on that for Eduardo, ping me if you want a verbal explanation before I get to it). 

Dividing by X is potentially a good idea since then we are more likely to flush closer to when a log has exceeded flush_ms. I'll change this, I'm gonna make X be 2.

File: plugins/filter_multiline/ml_concat.h
Comment (PettitWesley) [reply plugins/filter_multiline/ml_concat.h:77]: Good call. I keep forgetting about this. I wish C functions weren't in a global namespace...

File: plugins/filter_multiline/ml_concat.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml_concat.c:360]: Good call, that's safer.

File: plugins/filter_multiline/ml_concat.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml_concat.c:413]: AFAICT it does not. The buffer is the actual memory, the packer seems to be like a helper. The existing code for multiline has a packer and a buffer and it only frees the buffer. Also I tested this with valgrind and it didn't show a memory leak.

File: plugins/filter_multiline/ml_concat.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml_concat.c:166]: > packer->partial_id

This is an SDS string, so it is guaranteed null terminated.

File: plugins/filter_multiline/ml.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml.c:642]: I thought about this too myself... but returning directly from the filter is more efficient... also I think most of hte filters can be safely passed through twice... this is also something I can fix if it turns out it does have weird side effects and someone complains. 

There is an issue right now open that under high throughput the emitter fills up quickly... and remember the use case here is explicitly for huge logs, so I htink I want to stick with only using the emitter when needed.

File: plugins/filter_multiline/ml.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml.c:619]: Oh boy. This is actually a bug. Thanks. The key is not guaranteed to be "log"... it is user configured.

File: plugins/filter_multiline/ml.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml.c:656]: So the way this works is that you return the memory to the caller, and then the caller frees. If you look in the code in I think flb_filter.c it frees the returned buf. All of the filters work this way.

File: plugins/filter_multiline/ml.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml.c:520]: Check the beginning of cb_ml_filter:

```
    if (i_ins == ctx->ins_emitter) {
        flb_plg_trace(ctx->ins, "not processing records from the emitter");
        return FLB_FILTER_NOTOUCH;
    }
```

File: plugins/filter_multiline/ml.c
Comment (PettitWesley) [reply plugins/filter_multiline/ml.c:224]: Its stylistically better yes. Strictly speaking, if you fail plugin init, then Fluent Bit shuts down, and so freeing memory isn't needed. Lots of the plugins don't free on init failure.

**Comment by lubingfeng** [comment]
@edsiper  @matthewfala  are we okay to get this merged?

**Comment by lubingfeng** [comment]
@edsiper @koleini @fujimotos can we get this reviewed so that we can merge this? We get customers waiting for this fix.

**Comment by matthewfala** [comment]
@lubingfeng Thanks for checking. Let me take a second look.

**Review by matthewfala — APPROVED** [review]
The minor changes look good, and comments towards most of the other issues make sense!

I personally feel that the multiline function prefix should be something like flb_ml_concat_..., rather than just ml to follow the prefix convention set forth in flb_ml_parser.h/c, flb_ml_rule.h/c and other fluent bit files, and also to avoid naming collision with other libraries via the flb_ prefix. This is up to you and Eduardo though.

Not sure if you care about this style. But if you do, line 278 has an invisible extra indent in ml.c: https://github.com/fluent/fluent-bit/pull/5037/files#diff-d882420079c3f1dedc230e33a45fb81bf06a4deae9e2df11bd8ecaf509c73b0dR278

Lines 359 to 363 in ml_concat.c have some extra indents:
https://github.com/fluent/fluent-bit/pull/5037/files#diff-f5839e19406e810f65cbda8799c752423d8fb3c5ff50de0e220864d56d7b7c87R359-R363

Approving.

**Comment by matthewfala** [comment]
Approved with some optional change requests @lubingfeng

**Comment by PettitWesley** [comment]
@edsiper we actually should not merge this PR into 1.8 since 1.9 series has started and this is not a bug fix it is a new feature.  However I can not put up a PR against master for this change until this PR is merged: https://github.com/fluent/fluent-bit/pull/4671

**Comment by PettitWesley** [comment]
We will release this in 1.9 instead, and I will open a PR there once #4671 is merged

**Pull request closed without merging**
